### ISSUE TITLE
People: Face clusters missing after renaming and merging

### ISSUE THREAD

**Issue by pjft** [photoprism/photoprism#3414]
#### 1. What is not working as documented?

Renaming/merging face clusters can make faces no longer show up on Recognized people faces.

#### 2. How can we reproduce it?

1. Go to 'New' faces
2. Label people clusters, for instance "Manuel".
3. Mistakenly label one with "man" (had lots of clusters and was searching by name. Pressed "enter" assuming it would select "Manuel", but instead created a new cluster "man").
4. Click the "man" cluster label in that same screen, and rename it to "Manuel".

The "Manuel" cluster doesn't add these new clustered faces to it, and total stays the same.

Log messages are unexpected:

```
2023-05-17 22:18:20 INFO subject: person Manuel flagged as missing
2023-05-17 22:18:16 INFO changes successfully saved
2023-05-17 22:18:15 INFO subject: added person man
```

I expected "man" to be flagged as missing, not "Manuel".

Afterwards, I renamed "Manuel" to "man", to see if it changed anything, and then back to Manuel, to see if it'd absorb the old photos.

#### 3. What behavior do you expect?

The "Manuel" cluster adds the new clustered faces to it, showing up in the faces total. Also, "Manuel" shows up in the autocomplete suggestions.

#### 4. What could be the cause of your problem?

Potentially incorrectly flagging person as missing.

I tried both faces audit --fix and faces optimize, but nothing changed. I have a backup, and I suppose I could go directly to the database and try to fix things, but I'd rather not in case I break anything else.

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

This was when I made the mistake. I added "man", and then renamed to "Manuel". "Manuel" is flagged as missing, though it shouldn't have - there were already plenty of faces with that correct label:

```
2023-05-17 22:18:20 INFO subject: person Manuel flagged as missing
2023-05-17 22:18:16 INFO changes successfully saved
2023-05-17 22:18:15 INFO subject: added person man
```

Afterwards, whenever I try to add Manuel or man (to see if I could fix it?), I just get these:

-- adding man:
```
2023-05-17 23:32:14 INFO subject: person Manuel flagged as missing
2023-05-17 23:32:14 INFO subject: restoring person man
```

-- adding Manuel:
```
2023-05-17 23:31:27 INFO changes successfully saved
2023-05-17 23:31:26 INFO subject: person man flagged as missing
2023-05-17 23:31:26 INFO subject: restoring person Manuel
```

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: ARM64
(b) Database Type & Version: MariaDB
(c) Operating System Types & Versions: Linux Debian
(d) Browser Types & Versions: Chrome
(e) Ad Blockers, Browser Plugins, and/or Firewall Software? Yes.

Build 230514-0b780defb

#### 7. On what kind of device is PhotoPrism installed?

(a) Device / Processor Type: Raspberry Pi 4
(b) Physical Memory & Swap Space in GB: 8GB
(c) Storage Type: HDD
(d) Anything else that might be helpful to know? Not much.

**Comment by lastzero** [comment]
I just tested this on our demo and it seems to work as expected...

**Step 1**

![manuel1](https://github.com/photoprism/photoprism/assets/301686/f2bd332f-7910-4897-8372-5cdf3ff68e94)

**Step 2**

![manuel2](https://github.com/photoprism/photoprism/assets/301686/66baf7e1-06da-455c-9492-6e2515e33414)

**Result**

![manuel3](https://github.com/photoprism/photoprism/assets/301686/368ed937-f822-478e-8c9d-5fb9433eafa0)

**Please note:** The exact behavior will also depend on the specific faces due to conflict detection. In the worst case, e.g. if they may be identical twins, it could be that automatic recognition is disabled for specific faces as this was recently requested by a user.

**Comment by pjft** [comment]
Hi Michael!

You're right, there's the demo environment - perfect :)

Ok, I got it in that state right now fairly easily - if you go there (assuming it doesn't get recreated periodically), you'll find the moustached man and the Deutscher Filmpreis pink lady picture not showing in the face clusters.

What I did - after resetting all the faces last night by removing all the face mappings, and letting them re-cluster overnight, is the following:

-- always in the "New" People screen:
1. Click text box for man. Type "Manuel" and press enter.

```
2023-05-18 07:01:04 INFO subject: added person Manuel
2023-05-18 07:01:04 INFO changes successfully saved
```

2. Click text box for pink lady. Type "man". Press enter.

```
2023-05-18 07:01:07 INFO subject: added person man
2023-05-18 07:01:07 INFO changes successfully saved
```

3. Click "man" label, delete it and type Manuel. Press enter. Notice how it doesn't ask if we want to merge the clusters (just in case that's a symptom of the correct logic flow not being followed.

```
2023-05-18 07:01:13 INFO subject: person Manuel flagged as missing
2023-05-18 07:01:13 INFO changes successfully saved
```

Now, navigate to the Recognized screen to confirm only one cluster shows up. Refresh the page so total gets updated.

4. Click the "Manuel" label to edit it. Rename "Manuel" to man, to attempt to recover the pictures that may have been lost in the renaming. Cluster now disappears from view.

```
2023-05-18 07:01:41 INFO subject: restoring person man
2023-05-18 07:01:41 INFO subject: person man flagged as missing
```

Get scared. Refresh the page. Cluster is actually there. Phew!

5. Click the "man" label, and rename "man" to Manuel to get the actual name you intended back. Cluster disappears forever.

```
2023-05-18 07:01:54 INFO person saved
2023-05-18 07:01:54 INFO subject: person Manuel flagged as missing
```

So, what's the current state now:
- [If you search for "man"](https://demo.photoprism.app/library/browse?view=list&order=newest&q=man&public=true&quality=3), you will find the pictures for the "man" label face, even though it had supposedly been renamed Manuel. 
- If you open the picture, you will see the face labeled "man".
- The face label and cluster does not show up on the People tab.

If you can't replicate it I'll gladly reset it again and record the interaction, but hopefully this is helpful - and, assuming that it doesn't get recreated, you'll be able to check the state I got mine in.

I am thinking that I may fix it if I end up just manually removing the label from all the "Manuel" faces, one by one, and hoping that it clusters again, but since there are a lot of them, I was hoping that there'd be an easier way somehow. If not, well, it's ok. I'll just go through them one by one. In fact, I might, even though there are 700+ labels. 

If not, I have a database backup from before this happened that I can just restore.

Let me know if this helps, and thanks for the prompt response!

**Comment by pjft** [comment]
I just visited the demo environment and it was changed, so I assume it was regenerated or something was undone on the face clusters.

https://github.com/photoprism/photoprism/assets/22025709/77c55c64-2666-46e0-b89b-4b28ddc0133f

Still, I recorded a small video of me doing it again, in case it helps - with the logs window on the side.

Hope it helps.

**Comment by graciousgrey** [comment]
Thank you for the recording :)

**Comment by pjft** [comment]
By all means, thank you for the message!

For what it's worth, a few more notes:
- After manually going through the 700+ records of the face and "ejecting them" from the cluster, they finally all show up in the "New" face clusters.
- However, it seems that the label "Manuel" is still unusable. I tested adding a random face to "Manuel" and it still didn't show up as listed. Also, even though I labeled it "Manuel", it was automatically renamed as "man".

Hope this helps. I ended up confirming that there were no faces and markers tied to those 2 subject ids, and deleted them in the database.

I could also just restore the previous database backup but to be honest I'd rather avoid having to go tag my entire photo set if I can avoid it :)

**Comment by graciousgrey** [comment]
We will have a look at this next week.

**Comment by pjft** [comment]
Of course, no urgency whatsoever on my end, and thank you for engaging here - just wanted to report it in case it'd be helpful. I ended up recovering the previous database backup and decided to relabel the pictures again, so no problem there.

If there's anything I can do to help, though, let me know.

The only thing I noticed, if it helps, is that these labels still had a `deleted_at` value their entry in the `subjects`table, so it might be something as simple as seeing if that needs to be unset.

`select * from subjects where subj_name like 'man%' limit 5;`

was the query I was using to check that.

Have a great weekend! :)

**Comment by PhR34x0r** [comment]
Im having the same issue, not seeing two people because of a typo. Now their tags are mixed together and dont show up at all. Did you have a chance to look into this?
I tried "sudo docker compose exec photoprism photoprism faces update -f" but didnt help.
This really is a nasty bug with no way in the gui to fix it, appart from rescanning the entire library and basically starting over.
I have 60k photos and cant possibly look through all of them manually to remove the tags.

Are there commands i can run in the container to delete all data **of certain faces and labels** so they appear again in new and i can re-assign the same labels properly? Or is there any internal backup of the first finished scan before tagging, that i can go back to?

Thank you

**Comment by pjft** [comment]
My best bet, and always after making a backup of the database and all, would be to go straight to the database and try to fix it there. 

My last post stated where I ended there, but maybe it'll give you an idea of where to start looking.

Once again, this is not troubleshooting advice, pursuing this may result in your installation becoming corrupted, proceed with caution, etc.

**Comment by PhR34x0r** [comment]
Since i cant find people in an application for finding people, its already corrupted to me, doesnt matter if i break it any more. 
You obviously knew what you were doing, so if you have the time could you please share your steps, not as advice to do it, but so one doesnt have to start in the dark?
Guess the only other hope for us less savvy, is to wait for next development release where this is fixed, as learning sql from 0 likely takes longer.

**Comment by graciousgrey** [comment]
> Did you have a chance to look into this?

We haven't had time to provide a fix yet but it's on our list for the next release.

In the meantime, if you want to correct it yourself in the database, you can try the following:
- In the subjects table, you'll find a subject for each name you used (e.g. man and manuel in the case of the recording).
- For one of these subjects the photo_count is 0, for the other it is greater than 0 (the faces are assigned to this subject)
- Delete the entry of the subject with the photo_count = 0
- Set the field deletedAt of the subject whose photo_count is not 0 to null.

Now you should see the subject on the Recognized tab again and be able to rename it if needed. Note that this works for the scenario from the recording. If your case differs, you can share a database dump with us so that we can have a look at it.

In order to do a database backup before, you can follow this guide:
- https://docs.photoprism.app/user-guide/backups/

**Comment by pjft** [comment]
> Since i cant find people in an application for finding people, its already corrupted to me, doesnt matter if i break it any more. You obviously knew what you were doing, so if you have the time could you please share your steps, not as advice to do it, but so one doesnt have to start in the dark? Guess the only other hope for us less savvy, is to wait for next development release where this is fixed, as learning sql from 0 likely takes longer.

Alas, it was a while back, and trust me, I didn't know what I was doing - that's why I didn't attempt to fix it at all 😅 

Still, my notes for accessing the database were the following:

```
docker exec -it mariadb /bin/bash
mysql -uphotoprism -pinsecure
use photoprism
```

The subjects table referenced by @graciousgrey can be queried via

```
select * from subjects where subj_name like 'man%' limit 5;
```

I suspect you can then use a `delete` statement to delete the incorrect one, and an `update` statement to update the correct one, following the instructions provided moments ago.

I am not fully familiar with the mysql syntax nor the database schema, but I imagine that if you have an ID column you can just 

```
delete from subjects where id = <id of the row you want to update>
```

and 

```
UPDATE subjects
SET deletedAt = null
WHERE id = <id of the row you want to update>;
```

Hope this helps somewhat, but please backup your database beforehand. I have not tested any of this.

**Comment by PhR34x0r** [comment]
Thank you all for the quick replys, you're spot on. What I did:

docker compose exec photoprism photoprism backup -i -f
docker exec -it %actual mariadb containername% /bin/bash
mariadb -uphotoprism -p%actual password from compose yml file%
use photoprism
select * from subjects where subj_name like '%actual nametag%' limit 5;
delete from subjects where subj_slug = '%actual subj_slug%';
UPDATE subjects SET deleted_At = null WHERE subj_slug = '%actual subj_slug%';

Now one person with all the photos is back, but the two still get merged together and when i rename one picture in the stack, the entire stack gets renamed to the wrong person, instead of creating a new one for the right person.

Is there anything i can do about that?

edit: deleted the remaining name that had the pictures assigned, now all are gone. 
started an index scan, hoping both will show up as stack in the NEW tab, didnt happen.
started a new run of "photoprism update faces -f" now its all good. both people can be treated seperately again :)

thanks so much to both of you

**Comment by lastzero** [comment]
@pjft I've started a new development preview build to test these changes by either [pulling the preview image we provide](https://docs.photoprism.app/getting-started/updates/#development-preview) or [using our public demo](https://demo.photoprism.app/).

**Comment by pjft** [comment]
@lastzero thank you for the update! I'm currently away on annual leave, with minimal access to a proper laptop. I'll try it out on the public demo, and if it's still helpful in a few weeks when I have access to my machine I'll try to pull the preview image :)

**Comment by lastzero** [comment]
Note that these changes don't include database migrations to fix existing `subjects` tables. I would first need to get a better understanding how to do this reliably i.e. without unwanted side effects. Just removing the deleted flag from all rows would also show subjects that should be hidden.

**Comment by pjft** [comment]
Of course - didn't expect that yet. My database is currently ok, so I will not use this on it. I'll probably start from scratch or backup the database first just to try it out. I'll think about how to do the fix reliable as well. I imagine it isn't trivial to determine which are the ones that need to be fixed.

**Comment by pjft** [comment]
Hi @lastzero . Thanks for sending this over so far.

I tested it out and the good news is that the clusters don't disappear.

However, the behavior is still erratic. Basically, if I do the same sequences as before, there still seems to be some weird relationship or persistence between the two names ("Man" and "Manuel" or, in my new example, "Tim" and "Timothy").

https://github.com/photoprism/photoprism/assets/22025709/11e1c5be-f1fa-4474-8ab5-8feb925ceca1

Basically, after I merge "Tim" and "Timothy", next time I use the label that's not being used, it assumes I'm referring to the same entity and keeps swapping the names of everyone. See if the video helps somewhat.

**Comment by graciousgrey** [comment]
@pjft Thank you very much for testing :)

Could you try again with our latest preview, I tested with both sqlite and mariadb and could not reproduce your latest recording. For me everything worked as expected.

**Comment by pjft** [comment]
Hi @graciousgrey ! Thank you for the heads up - I will do my best to go through it this week and report back. :)

**Comment by pjft** [comment]
Hi @graciousgrey . Thanks for the heads up once again.

From my limited testing, I also think it is now working as intended. Thank you for looking into this!

I am happy to close this unless you still have anything else you want to do that needs to keep this open.
